File: Source/Website/sitecore/shell/client/YourApps/WorkflowDataSource/WorkflowDatasource.js
define(["sitecore"], function (Sitecore) {
    var model = Sitecore.Definitions.Models.ControlModel.extend({
        initialize: function (options) {
            this._super();

            this.set("BeginDate", "");
            this.set("EndDate", "");
            this.set("Output", "");
            this.set("errorText", "");

            //Subscribe to Begin and End date changes
            this.on("change:BeginDate", this.updateItems, this);
            this.on("change:EndDate", this.updateItems, this);
        },

        convertToDate: function (date) {
            /// <summary>
            ///     Converts the given string date into a Date object
            /// </summary>
            var matches = /^(\d{1,2})[-\/](\d{1,2})[-\/](\d{4})$/.exec(date);
            if (matches == null) return undefined;
            var d = matches[2];
            var m = matches[1] - 1;
            var y = matches[3];
            var composedDate = new Date(y, m, d);
            return composedDate;
        },

        isValidDate: function (date) {
            /// <summary>
            ///     Ensures the date string provided is an actual date
            /// </summary>
            var matches = /^(\d{1,2})[-\/](\d{1,2})[-\/](\d{4})$/.exec(date);
            if (matches == null) return false;
            var d = matches[2];
            var m = matches[1] - 1;
            var y = matches[3];
            var composedDate = new Date(y, m, d);
            return composedDate.getDate() == d &&
                    composedDate.getMonth() == m &&
                    composedDate.getFullYear() == y;
        },

        beginDateBeforeEndDate: function (beginDateStr, endDateStr) {
            /// <summary>
            ///     Ensures the begin date occurs befor the end date
            /// </summary>
            var beginDate = this.convertToDate(beginDateStr);
            var endDate = this.convertToDate(endDateStr);

            return endDate > beginDate;
        },

        updateItems: function () {
            /// <summary>
            ///     Updates the output using the given input using ajax
            /// </summary>

            //Grab the begin and end date
            var beginDate = this.get("BeginDate");
            var endDate = this.get("EndDate");

            //Perform validations on the dates to ensure that they are properly formatted and aligned
            if (this.isValidDate(beginDate) && this.isValidDate(endDate) && this.beginDateBeforeEndDate(beginDate, endDate)) {

                //Do an ajax call to the workflow data source item to retrieve a list of information regarding workflow between the two dates
                $.ajax({
                    url: "/api/sitecore/WorkflowDatasource/GetWorkflowData",
                    type: "POST",
                    data: {
                        beginDate: beginDate,
                        endDate: endDate
                    },
                    context: this,
                    success: function (data) {
                        var jsonData = JSON.parse(data);
                        if (jsonData.success) {
                            this.set("errorText", "");
                            this.set("Output", jsonData.data);
                        }
                        else {
                            this.set("errorText", jsonData.errorText);
                        }
                    }
                });
            }
        },

    });

    var view = Sitecore.Definitions.Views.ControlView.extend({
        initialize: function (options) {
            this._super();

        }
    });

    Sitecore.Factories.createComponent("WorkflowDatasource", model, view, ".sc-WorkflowDatasource");
});

File: Source/Website/sitecore/shell/client/YourApps/WorkflowDataSource/WorkflowDatasource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using System.Web.Script.Serialization;
using Sitecore.Configuration;
using Sitecore.ContentSearch.SearchTypes;
using Sitecore.ContentSearch.Security;
using Sitecore.Data.Items;
using Sitecore.Diagnostics;
using Sitecore.Web;
using Sitecore.ContentSearch.Linq;
using Sitecore.Workflows;

namespace Heisenberg.sitecore.shell.client.YourApps.WorkflowDataSource
{
    public class WorkflowDatasourceController : Controller
    {
        /// <summary>
        ///     Accepts "BeginDate" and "EndDate" and retrieves all workflow records within that time span, one per item
        ///     Returns a JSON formatted string of:
        ///     {
        ///         errorMessage: string,
        ///         success: bool,
        ///         data: [UserName, IssuesUpdated]
        ///     }
        /// </summary>
        /// <returns></returns>
        public string GetWorkflowData()
        {
            //Initalize possible error message to be returned
            string errorMessage = string.Empty;

            //Grab begin and end date from the form data (Passed in as data using ajax)
            var beginDate = WebUtil.GetFormValue("BeginDate");
            var endDate = WebUtil.GetFormValue("EndDate");

            //Initalize return collection
            var data = new List<WorkflowUserRecords>();

            //Initalize Date Types
            DateTime begin;
            DateTime end;

            //Server side validation of date range
            if (DateTime.TryParse(beginDate, out begin) && DateTime.TryParse(endDate, out end) && begin < end)
            {
                //Add a day to end date to ensure we're getting changes made late into the selected day
                end = end.AddDays(1);

                //Grab the master indexer
                var index = Sitecore.ContentSearch.ContentSearchManager.GetIndex("sitecore_master_index");
                //Grab the master database
                var masterDb = Factory.GetDatabase("master");

                //Using the search context for the master indexer
                using (var searchIndex = index.CreateSearchContext(SearchSecurityOptions.DisableSecurityCheck))
                {
                    //Retrieve all items that have had their Updated date within the date range provided
                    var results =
                        searchIndex.GetQueryable<SearchResultItem>()
                            .Where(i => i.Updated > begin && i.Updated < end);

                    //Grab the SearchResultItems as actual sitecore items
                    var items = results.Select(i => i.GetItem());

                    //Get the workflow provider from the master database
                    var workflowProvider = masterDb.WorkflowProvider;

                    //Loop over the filtered items list
                    foreach (var item in items)
                    {
                        WorkflowEvent lastHistoryItem;

                        //Gets the latest workflow history entry for the current item. Returns false if any issues are found while getting entry
                        if (GetLatestWorkflowHistoryEntry(workflowProvider, item, begin, end, out lastHistoryItem))
                            continue;

                        //Update the data for the user that made the last history entry
                        UpdateDataForUser(data, lastHistoryItem);
                    }
                }
            }
            else
            {
                errorMessage = "Entries given are not in the correct date format.";
            }

            var returnValues = new
            {
                data,
                errorMessage,
                success = string.IsNullOrEmpty(errorMessage)
            };

            return new JavaScriptSerializer().Serialize(returnValues);
        }

        /// <summary>
        ///     Ensures a record exists for a user before adding to the list of issues they worked
        /// </summary>
        /// <param name="data"></param>
        /// <param name="lastHistoryItem"></param>
        private static void UpdateDataForUser(List<WorkflowUserRecords> data, WorkflowEvent lastHistoryItem)
        {
            //Retrieve the user that made the update from our return data, or null if no entry exist.
            //This acts similarly to a dictionary but gives us control of more information to be added later
            var person = data.FirstOrDefault(dataItem => dataItem.UserName == lastHistoryItem.User);

            //If the user that made the change to the current item has no previous entry make one
            if (person == null)
            {
                //Create entry for the user and add them to the return data
                person = new WorkflowUserRecords
                {
                    UserName = lastHistoryItem.User,
                    IssuesUpdated = 0
                };
                data.Add(person);
            }
            //Update the number of issues updated for the returned user
            person.IssuesUpdated++;
        }

        /// <summary>
        ///     Retrieves the latest history entry and determines if its latest modify date is within the date range
        /// </summary>
        /// <param name="workflowProvider"></param>
        /// <param name="item"></param>
        /// <param name="begin"></param>
        /// <param name="end"></param>
        /// <param name="lastHistoryItem"></param>
        /// <returns></returns>
        private static bool GetLatestWorkflowHistoryEntry(IWorkflowProvider workflowProvider, Item item, DateTime begin,
            DateTime end, out WorkflowEvent lastHistoryItem)
        {
            lastHistoryItem = null;

            //Get the workflow for the current item
            var itemWorkflow = workflowProvider.GetWorkflow(item);

            //If the workflow for the item is null then move to the next item
            if (itemWorkflow == null) return true;
            //Grab the history of the items workflow
            var itemHistory = itemWorkflow.GetHistory(item);

            //If the history of the item is null or has no entries then move to the next item
            if (itemHistory == null || itemHistory.Length <= 0) return true;

            //Grab the latest history update to an item
            lastHistoryItem = itemHistory[itemHistory.Length -1];

            //Ensure that the latest history update exist in the specified date range
            return lastHistoryItem.Date <= begin || lastHistoryItem.Date >= end;
        }

        /// <summary>
        ///     Simple Data Object for simple JSON seralizing
        /// </summary>
        private class WorkflowUserRecords
        {
            public string UserName { get; set; }
            public int IssuesUpdated { get; set; }
        }
    }

}